lib/puppet/type.rb

```diff
@@ -107,13 +107,14 @@ class << self
   end
 
   # Returns all the attribute names of the type in the appropriate order.
-  # The {key_attributes} come first, then the {provider}, then the {properties}, and finally
-  # the {parameters} and {metaparams},
+  # The {key_attributes} come first, then the {provider} and {checksum}, then the {properties},
+  # and finally the {parameters} and {metaparams},
   # all in the order they were specified in the respective files.
   # @return [Array<String>] all type attribute names in a defined order.
   #
   def self.allattrs
-    key_attributes | (parameters & [:provider]) | properties.collect { |property| property.name } | parameters | metaparams
+    # Checksum has no side-effects, and should be set early for properties (like content) that use it.
+    key_attributes | (parameters & [:provider, :checksum]) | properties.collect { |property| property.name } | parameters | metaparams
```

Comment:
Oh, as in external types. Ok, I'll just drop that commit.

On Fri, Jan 30, 2015 at 3:20 PM, Josh Cooper <notifications@github.com>
wrote:

> In lib/puppet/type.rb
> <https://github.com/puppetlabs/puppet/pull/3541#discussion_r23880134>:
>
> >    # all in the order they were specified in the respective files.
> >    # @return [Array<String>] all type attribute names in a defined order.
> >    #
> >    def self.allattrs
> > -    key_attributes | (parameters & [:provider]) | properties.collect { |property| property.name } | parameters | metaparams
> > +    # Checksum has no side-effects, and should be set early for properties (like content) that use it.
> > +    key_attributes | (parameters & [:provider, :checksum]) | properties.collect { |property| property.name } | parameters | metaparams
>
> I forgot this was in type.rb. I don't think we can do this here, as who
> knows what other types have been created with a checksum parameter.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/puppet/pull/3541/files#r23880134>.
>